[충남대 FE 김은수] 2단계 선물하기 메인 페이지 UI 구현하기

shushuburger opened the pull request:
멘토님 안녕하세요. 충남대학교 FE12팀 김은수입니다.

## 과제 요구사항

- [x] 본인만의 기준으로 일관된 코드를 작성해주세요.
- [x] 기능 단위로 나누어 커밋을 해주세요.
- [x] emotion에 Color, Typography 토큰을 설정해요. [(토큰 가이드)](https://edu.nextstep.camp/s/0eoNzeZS/ls/lNeUqOll)
> src/styles/tokens에 color.ts, spacing.ts, typography.ts를 각각 작성하여 theme.ts로 묶어 export 해주었습니다.
> 이후 components들에서 theme을 기반으로 스타일 적용을 해주었습니다.
- [x] Mobile First Design 기반의 Layout을 구현해요. (아래 설명 참고)
> 화면의 max-width를 720px으로 제한하였습니다.
- [x] 상단 네비게이션 바 영역을 구현해요.
> Navigation.tsx에 구현하였습니다.
>> 스크롤 시 네비게이션 바는 상단에 고정되어 있도록 해두었습니다.
- [x] 카테고리 섹션을 구현해요.
> 주어진 mock data를 src/data에 categories.ts로 저장하여 활용하였습니다.
> CategorySection.tsx에 구현하였습니다.
- [x] 기타 섹션 (선물한 친구 선택, 카테캠 화이팅) UI를 구현해요.
> 각각 친구 선택은 FriendSelectSection.tsx, 카테캠 화이팅은 BannerSection.tsx에 구현하였습니다.
- [x] 실시간 급상승 선물랭킹을 구현해요.
> 주어진 mock data를 src/data에 products.ts로 저장하여 활용하였습니다.
> RankingSection.tsx에 구현하였습니다.

## 참고 링크
- https://meyerweb.com/eric/tools/css/reset/ : reset.css 적용 시에 참고한 링크입니다.
- https://emotion.sh/docs/%40emotion/css : emotion을 사용할 때 참고한 링크입니다.
- https://carbondesignsystem.com/elements/color/overview/ : token을 작성하고 그를 이용하기 위해 코드 작성 시에 참고한 링크입니다.

shushuburger committed:
docs(readme): 2단계 과제 시작 전 구현해야 하는 목록 작성

- README에 구현해야 하는 것들 목록을 작성했습니다.
- index.html에서 문서의 언어를 한국어로 변경해주었습니다.

shushuburger committed:
feat(style): Emotion 기반 reset 및 base 스타일 분리 적용

- reset.css 공식 문서를 바탕으로 ResetStyle.tsx를 작성해주었습니다.
- 1차 과제의 조건이었던 font 적용을 위해 BaseStyle.tsx를 통해 기본적인 스타일을 작성했습니다.
- Emotion 사용을 위해 GlobalStyle.tsx를 통해 ResetStyle과 BaseStyle을 모두 묶어 export 해주었습니다.

shushuburger committed:
feat(theme): 디자인 토큰 정의 및 theme 적용

- color.ts, spacing.ts, typography.ts 파일로 토큰 분리 및 as const 적용하였습니다.
- theme.ts 파일 생성하여 토큰들을 한 번에 관리할 수 있도록 하였습니다.
- emotion.d.ts 작성하여 Emotion Theme에 MyTheme 타입 연결해 주었습니다.
- BaseStyle에서 theme 사용하도록 코드를 수정하였습니다.
- GlobalStyle에서 useTheme 사용하여 theme를 BaseStyle의 인자로 전달하도록 해주었습니다.
- 정상적으로 token을 이용한 스타일이 반영됨을 확인하였습니다.
- README에 구현 완료를 표시했습니다.

shushuburger committed:
feat(layout): mobile first design 구현

- mobile first design을 위한 요구사항인 max-width: 720px를 적용했습니다.
- width: 100%와 margin: 0 auto로 가운데 정렬 설정을 완료했습니다.
- README에 구현 완료를 표시했습니다.

shushuburger committed:
feat(data): mock data추가

- categories.ts와 products.ts에 각각 주어진 mock data 삽입했습니다.
- 다른 컴포넌트에서 import 가능하도록 export 처리해 두었습니다.

shushuburger committed:
feat(component): Navigation 컴포넌트 구현 및 토큰 기반 스타일 적용

- 좌측 뒤로가기, 중앙 로고, 우측 로그인 버튼으로 네비게이션 바를 구성했습니다.
- spacing, color 토큰을 활용하여 Emotion 스타일을 정의하였습니다.
- present.webp 이미지를 src/assets에서 import 방식으로 사용할 수 있게 하였습니다.
- 로고에 커서 포인터 추가하여 클릭 가능한 것처럼 보이도록 해두었습니다.
- README에 구현 완료를 표시했습니다.

shushuburger committed:
feat(component): FriendSelect 컴포넌트 추가 및 스타일 구성

- "선물할 친구를 선택해 주세요." 영역에 대한 UI 구현하였습니다.
- 참고 링크와 비교하며 유사한 디자인을 가지도록 수정해주었습니다.

shushuburger committed:
feat(component): CategorySection 컴포넌트 구현

- categories.ts 데이터를 기반으로 카테고리 UI를 구현하였습니다.
- Grid를 이용해 한 줄에 5개의 아이템이 일관된 형태로 뜨도록 해주었습니다.
- theme을 적용하여 스타일 작업을 해주었습니다.
- README에 구현 완료를 표시하였습니다.

shushuburger committed:
feat(component): BannerSection 컴포넌트 구현

- 프론트엔드 2단계 과제를 응원 배너 UI를 추가했습니다.
- color.ts, typography.ts, spacing.ts에 있는 내용들로 스타일을 적용했습니다.
- README에 구현 완료를 표시했습니다.

shushuburger committed:
feat(ranking): 필터/정렬 UI 상태 관리 및 스타일 적용

- filter 버튼 클릭 시 선택된 상태를 useState로 관리하도록 했습니다.
- 선택된 필터에 따라 emoji, label 스타일 동적 적용되도록 했습니다.
- 더보기/접기 버튼으로 리스트 확장하고 축소할 수 있도록 했습니다.
- README에 구현 완료를 표시했습니다.

shushuburger committed:
feat(navigation): 상단 네비게이션 스크롤 시 고정되도록 sticky 스타일 적용

- Navigation 컴포넌트에 position: sticky 및 top: 0 설정해주었습니다.
- 스크롤 시 항상 네비게이션이 상단에 고정되도록 개선하였습니다.

shushuburger committed:
refactor(theme): 스타일을 theme 단축 표현으로 수정

- theme.typography를 활용해 font-size, font-weight, line-height 등을 한 줄로 축약하였습니다.
- 전체 컴포넌트에 theme에 기반한 스타일 적용을 진행하였습니다.

shushuburger committed:
style(reset): 명명 규칙 정리 및 불필요한 padding 초기화 제거

- 스타일 변수 및 CSS 선택자 명명 일관성을 한 번더 점검하며 개선했습니다.
- ResetStyle 내 사용되지 않는 padding: 0 선언을 제거했습니다.

shushuburger committed:
docs(readme): 프로젝트 구현 완료 기록

- 과제 요구사항을 모두 수행했음을 표시했습니다.

igy95 commented on src/assets/present.webp:
png, webp 확장자를 가진 에셋들이 보이는데, 확장자를 관리하는 기준이 따로 있으신지 궁금합니다!

shushuburger replied:
아니요! 딱히 확장자에 관한 기준을 삼고 있지는 않습니다.
다만 https://developer.mozilla.org/en-US/docs/Web/Media/Guides/Formats/Image_types#webp 를 바탕으로 해서 확장자를 정할 때 webp, png를 가장 우선순위에 두고 있습니다.

추가적으로 찾아보니 https://cloudconvert.com/png-to-webp 와 같이 webp로 확장자를 변환할 수 있는 사이트가 있기에 앞으로 webp를 최우선으로 해서 작성해보려고 합니다.

igy95 commented on src/components/BannerSection.tsx:
```diff
@@ -0,0 +1,40 @@
+const BannerSection = () => {
+  const theme = useTheme();
+
+  const sectionStyle = css`
+    padding: ${theme.spacing[0]} ${theme.spacing[4]};
+    margin-bottom: ${theme.spacing[8]};
+  `;
+
+  const textWrapperStyle = css`
+    display: flex;
+    flex-direction: column;
+    background-color: ${theme.color.semantic.kakaoYellow};
+    border-radius: 16px;
+    padding: ${theme.spacing[4]};
+  `;
+
+  const topTextStyle = css`
+    ${theme.typography.label.label2Regular};
+    color: ${theme.color.gray[700]};
+  `;
+
+  const bottomTextStyle = css`
+    ${theme.typography.label.label1Bold};
+    color: ${theme.color.semantic.text.default};
+  `;
```
theme 때문에 스타일 코드가 컴포넌트 안으로 들어오게 된 것 같은데, 

1. theme을 가져오는 방식이 useTheme밖에 없었는지 확인해보시면 좋겠습니다
2. 만일 useTheme 사용이 어쩔 수 없다면, 스타일 코드에서 인자를 받도록 열어두고 이건 컴포넌트 외부에 선언해두는 게 좋을 것 같아요!

컴포넌트 내부에 jsx, 훅, 비즈니스 로직 외에 스타일 코드까지 모두 가져버리면 코드를 확인하기도 어려워지고 컴포넌트가 호출될 때마다 모두 재선언되는 비용이 들기 때문이에요!

shushuburger replied:
1. https://emotion.sh/docs/theming 해당 링크에서 styled component 방식을 습득하였습니다.

해당 방식으로 변경 완료하였습니다!

igy95 commented on src/components/CategorySection.tsx:
emotion/react의 styled component 방식 대신 css 합성 방식을 일관적으로 가져가신 이유가 있을까요??

shushuburger replied:
특별한 이유는 없습니다.
https://emotion.sh/docs/theming
해당 링크를 참고해서 theme을 구성하였고 가장 위에 보인 것이 css 합성 방식이었기에 이 방식을 선택했었습니다.

그렇지만 위에 주신 "theme을 useTheme 말고 사용할 수 있는 다른 방법"에 대한 방식을 찾다가 

```
 const SomeText = styled.div`
  color: ${props => props.theme.colors.primary};
`
```

styled component라면 useTheme 없이도 theme을 사용할 수 있다라는 것을 알게 되었습니다.
이에 따라 기존의 css 사용 방식을 styled component 방식으로 변경할 예정입니다!

igy95 commented on src/components/Navigation.tsx:
```diff
@@ -0,0 +1,81 @@
+      </div>
+
+      <div>
+        <button css={iconButtonStyle} aria-label="로그인">
```
aria-label은 어떤 의도로 작성해주셨을까요~?

shushuburger replied:
스크린 리더 사용자를 고려하여 코드를 작성해둔 것입니다!

현재 버튼이 아이콘으로만 구성되어 있는 상황이어서 스크린 리더 사용자는 버튼의 텍스트 설명이 없으면 무슨 역할인지 알 수 없습니다.
그래서 스크린 리더 사용자 또한 이 버튼이 어떠한 역할을 하는지 알수 있게 해주기 위해 aria-label을 사용했습니다.

해당 내용을 참고하였습니다.
https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-label

igy95 commented on src/components/RankingSection.tsx:
```diff
@@ -0,0 +1,238 @@
+
+const RankingSection = () => {
+  const theme = useTheme();
+  const [visibleCount, setVisibleCount] = useState(6);
```
6이라는 값은 매직 넘버로 두기보다는 별도의 상수로 관리하는 게 좋겠습니다!

shushuburger replied:
`const INITIAL_VISIBLE_COUNT = 6;
`

을 통해 상수로 관리하도록 변경하였습니다!

igy95 commented on src/components/RankingSection.tsx:
이 컴포넌트는 한 눈에 봐도 꽤 큰 사이즈로 보이는데요! 섹션 안에서도 컴포넌트를 더 분리해볼 수 있지 않을까요??

shushuburger replied:
RankingSection이라는 폴더를 만든 다음
- 성별 필터, 우선순위 필터, 상품 목록 화면, 상품 개별 컴포넌트 등

으로 나누어 분리하였습니다!

igy95 commented on src/data/categories.ts:
```diff
@@ -0,0 +1,98 @@
+export interface Category {
+  themeId: number;
+  name: string;
+  image: string;
+}
```
타입은 다른 곳에서도 공유될 수 있기 때문에 다른 파일에 따로 선언해두는 게 좋겠습니다 !

shushuburger replied:
- src/types/category.ts
- src/types/product.ts

이렇게 2개의 파일을 만들어 기존 src/data 내의 categories.ts, products.ts는 import를 통해 타입을 사용하는 방식으로 변경하였습니다.

igy95 commented on src/styles/BaseStyle.tsx:
```diff
@@ -0,0 +1,27 @@
+import { css } from '@emotion/react';
+import type { Theme } from '@/styles/theme';
+
+export const BaseStyle = (theme: Theme) => css`
```
style component 방식으로 작성하셨다면 파스칼 케이스가 자연스럽지만, 단순 css 스타일을 반환하는 변수의 이름은 카멜 케이스가 적절할 것 같습니다. 리액트 맥락에서 파스칼 케이스는 대부분 타입 혹은 컴포넌트로 읽히기 때문이에요!

shushuburger replied:
ResetStyle.ts와 BaseStyle.ts에서 각각
- ResetStyle -> resetStyle
- BaseStyle -> baseStyle

로 변경해준 다음 GlobalStyle.tsx에서도 변경을 진행해주었습니다.

igy95 commented on src/styles/ResetStyle.tsx:
파일 확장자명이 tsx여야할 필요가 있을까 싶네요!

shushuburger replied:
ResetStyle.ts, BaseStyle.ts로 수정 완료하였습니다!

igy95 reviewed (changes requested):
안녕하세요 은수님! 이번 단계도 고생 많으셨습니다 ㅎㅎ 리뷰 코멘트 확인 후 재요청 부탁드려요! 그리고 개발하시면서 궁금한 부분이 있다면 본문이나 코멘트에 얼마든 질문 주셔도 좋습니다 ~~

shushuburger committed:
refactor(style): BaseStyle.ts 및 resetStyle.ts 네이밍 및 확장자 정리

- BaseStyle.tsx → baseStyle.ts로 파일명 변경 및 변수명을 카멜 케이스로 수정하였습니다.
- ResetStyle.tsx → resetStyle.ts로 파일명 변경 및 변수명을 카멜 케이스로 수정하였습니다.

shushuburger committed:
refactor(types): category 및 product 타입 분리 및 파일 정리

- Category, Product 인터페이스를 각각 types/category.ts, types/product.ts로 분리하였습니다.
- mockCategories, mockProducts는 기존 위치 유지하며 타입 import 방식으로 변경하였습니다.

shushuburger committed:
refactor(ranking): 초기 노출 상품 개수 상수로 분리

- visibleCount 초기값으로 사용되던 숫자 6을 INITIAL_VISIBLE_COUNT 상수로 추출하도록 했습니다.
- toggleVisibleCount 로직 내에서도 6 대신 해당 상수를 사용하도록 했습니다.

shushuburger committed:
refactor(ranking): RankingSection을 역할별 컴포넌트로 분리

- RankingSection 내 UI 요소들을 기능별로 분리하여 구조를 개선하였습니다.
- RankingFilter, RankingSort, ProductGrid, ProductCard, ExpandButton 컴포넌트로 세분화하여 저장했습니다.

shushuburger commented:
궁금한 점이 생겼습니다!!

1) 각 componenet 파일 코드에는 style 관련 코드, render 관련 코드가 존재합니다. 
저는 style부터 작성하고 하단에 render에 대한 코드를 작성하는 방식을 사용했습니다.
> 공식 사이트 (https://emotion.sh/docs/theming) 에서는 "theme에 관한 코드를 먼저 작성 -> render를 마지막에" 하기도 하고 "render -> props" 방식을 사용하기도 했습니다. 주로 "theme -> render" 였습니다.
그러던 중 ai를 이용해 css와 styled component의 차이에 대해 더 알아보던 중 ai가 "render -> styled component" 순으로 예제 코드를 주는 것을 보았습니다.
그 이유를 물어보니 "가독성 향상"이 가장 큰 이유라고 했습니다.

공식에서 순서를 혼용했던 이유는 예제 코드가 짧아서였을까요? 
저는 코드가 길어질수록 "render -> style" 방식이 더 좋을 것 같다고 생각이 듭니다. 
멘토님께서는 어떤 방식을 더 선호하시는지 궁금해졌습니다!



++++

### 아 그리고 만약에 아직 코드 리뷰를 다 반영하지 않아서 "리뷰 요청"을 드리기에는 이른 시점이지만 작업 중 멘토님께 여쭤보고 싶은 것이 생긴다면 어떻게 하면 될까요?
- github에 궁금증을 제가 comment로 달기만 하면 멘토님께 알림이 가나요? 그게 아니라면 질문을 작성하고 어떻게 멘토님께 전달할 수 있는지 알려주시면 감사하겠습니다!!

shushuburger committed:
refactor(style): Emotion css → styled-components 방식으로 수정

- 기존 useTheme + css 함수 방식 제거하였습니다.
- 모든 컴포넌트를 styled-components 방식으로 통일하였습니다.

igy95 commented:
> 저는 코드가 길어질수록 "render -> style" 방식이 더 좋을 것 같다고 생각이 듭니다. 멘토님께서는 어떤 방식을 더 선호하시는지 궁금해졌습니다!

아무래도 style 코드가 길어지다보면 컴포넌트 코드를 바로 확인할 수 없기 때문에 render 코드를 상단에 배치시키는 걸 권장한다고 생각합니다. 저도 이러한 측면에서 컴포넌트 파일과 스타일 파일을 나누거나 (e.g. component.tsx, component.styled.ts) 말씀 주신 것처럼 한 파일 내에서 컴포넌트 코드를 우선으로 두면 될 것 같습니다 !

> 아 그리고 만약에 아직 코드 리뷰를 다 반영하지 않아서 "리뷰 요청"을 드리기에는 이른 시점이지만 작업 중 멘토님께 여쭤보고 싶은 것이 생긴다면 어떻게 하면 될까요?

그럴 경우 디스코드 dm으로 주시면 제가 확인하도록 하겠습니다 !

igy95 reviewed (approved):
안녕하세요 은수님! 이번 리뷰가 늦어 죄송합니다 🙏 코멘트 남겨주신 부분을 토대로 어떤 방향으로 코드를 작성하려고 하셨는지, 어떻게 수정하셨는지 잘 공유해주셔서 반영 사항 확인이 수월했습니다 :) 추가 코멘트 남겨두었으니 확인 후 다음 단계에서 뵐게요 ~
